File: 173_Binary_Search_Tree_Iterator.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */
public class BSTIterator {
    private Stack<TreeNode> s;
    public BSTIterator(TreeNode root) {
        s = new Stack<TreeNode>();
        PushAllLefts(root);
    }
    
    /** @return the next smallest number */
    public int Next() {
        var node = s.Pop();
        var cur = node.right;
        if (cur != null) {
            PushAllLefts(cur);
        }
        return node.val;
    }
    
    /** @return whether we have a next smallest number */
    public bool HasNext() {
        return s.Count() > 0;
    }
    
    private void PushAllLefts(TreeNode node) {
        for (; node != null; node = node.left)
            s.Push(node);
    }
}

/**
 * Your BSTIterator object will be instantiated and called as such:
 * BSTIterator obj = new BSTIterator(root);
 * int param_1 = obj.Next();
 * bool param_2 = obj.HasNext();
 */
 

File: 173_Binary_Search_Tree_Iterator.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class BSTIterator:

    def __init__(self, root: TreeNode):
        self.stack = []
        self.appendLeftSubtree(root)

    def next(self) -> int:
        """
        @return the next smallest number
        """
        ret = self.stack.pop()
        if ret.right is not None:
            self.appendLeftSubtree(ret.right)
        
        return ret.val

    def hasNext(self) -> bool:
        """
        @return whether we have a next smallest number
        """
        return len(self.stack) > 0
    
    def appendLeftSubtree(self, root):
        cur = root
        while cur is not None:
            self.stack.append(cur)
            cur = cur.left
        return


# Your BSTIterator object will be instantiated and called as such:
# obj = BSTIterator(root)
# param_1 = obj.next()
# param_2 = obj.hasNext()
